src/compiler/crystal/semantic/abstract_def_checker.cr

```diff
@@ -276,6 +281,45 @@ class Crystal::AbstractDefChecker
     true
   end
 
+  def same_parameters?(m1 : Def, m2 : Def)
+    return false unless m1.args.size == m2.args.size
+
+    splat_index = m1.splat_index
+    return false unless splat_index == m2.splat_index
+
+    named_args1 = nil
+    named_args2 = nil
```

Comment:
It's a performance optimization to avoid the (double) array allocation for the (more likely) case that there are no named args. So I think the current code is probably better (although perhaps not as elegant).

Reply:
However, we can know beforehand if there *could* be an allocation for named args (only if `args.size > splat_index`).

And I suppose the loop could be refactored and split into different stages:
```cr
(0...splat_index || m1.args.size).times do |i|
  arg1 = m1.args[i]
  arg2 = m2.args[i]
  return false unless arg1.external_name == arg2.external_name
end

if splat_index
  return false unless (m1.args[splat_index].external_name != "") == (m2.args[splat_index].external_name != "")

  named_args1 = m1.args[splat_index..].sort!
  named_args2 = m2.args[splat_index..].sort!

  return false unless named_args1 == named_args2
end

m1.double_splat.nil? == m2.double_splat.nil?
```
I think this should be semantically equivalent (more or less, haven't tested it). And it's definitely more concise. Not sure about performance characteristics, but I doubt it makes a huge difference.